src/catalog/dependency_manager.cpp

```diff
@@ -27,166 +115,311 @@ void DependencyManager::AddObject(CatalogTransaction transaction, CatalogEntry &
-		auto &set = dependents_map[dependency];
-		set.insert(Dependency(object, dependency_type));
+		auto &dependency_connections = GetOrCreateDependencySet(transaction, dependency);
+		dependency_connections.AddDependent(transaction, object, dependency_type);
 	}
 	// create the dependents map for this object: it starts out empty
-	dependents_map[object] = dependency_set_t();
-	dependencies_map[object] = dependencies.set;
+
+	auto &object_connections = GetOrCreateDependencySet(transaction, object);
+	object_connections.AddDependencies(transaction, dependencies);
+}
+
+static bool CascadeDrop(bool cascade, DependencyType dependency_type) {
+	if (cascade) {
+		return true;
+	}
+	if (dependency_type == DependencyType::DEPENDENCY_AUTOMATIC) {
+		// These dependencies are automatically dropped implicitly
+		return true;
+	}
+	if (dependency_type == DependencyType::DEPENDENCY_OWNS) {
+		// The object has explicit ownership over the dependency
+		return true;
+	}
+	return false;
+}
+
+void DependencyManager::UnmangleName(const string &mangled, CatalogType &type, string &schema, string &name) {
+	auto parts = StringUtil::Split(mangled, std::string("\0", 1));
+	D_ASSERT(parts.size() == 3);
+	type = CatalogTypeFromString(parts[0]);
+	schema = std::move(parts[1]);
+	name = std::move(parts[2]);
+}
+
+void GetLookupProperties(CatalogEntry &entry, string &schema, string &name, CatalogType &type) {
+	if (entry.type == CatalogType::DEPENDENCY_ENTRY) {
+		auto &dependency_entry = entry.Cast<DependencyCatalogEntry>();
+
+		schema = dependency_entry.EntrySchema();
+		name = dependency_entry.EntryName();
+		type = dependency_entry.EntryType();
+	} else if (entry.type == CatalogType::DEPENDENCY_SET) {
+		auto &dependency_set = entry.Cast<DependencySetCatalogEntry>();
+
+		auto &mangled_name = dependency_set.MangledName();
+		DependencyManager::UnmangleName(mangled_name, type, schema, name);
+	} else {
+		throw InternalException("Unrecognized CatalogType in 'GetLookupProperties'");
+	}
+}
+
+// Always performs the callback, it's up to the callback to determine what to do based on the lookup result
+optional_ptr<CatalogEntry> DependencyManager::LookupEntry(CatalogTransaction transaction, CatalogEntry &dependency) {
+	string schema;
+	string name;
+	CatalogType type;
+	GetLookupProperties(dependency, schema, name, type);
+
+	// Lookup the schema
+	auto schema_entry = catalog.schemas->GetEntry(transaction, schema);
+	if (type == CatalogType::SCHEMA_ENTRY || !schema_entry) {
+		// This is a schema entry, perform the callback only providing the schema
+		return schema_entry;
+	}
+	auto &duck_schema_entry = schema_entry->Cast<DuckSchemaEntry>();
+
+	// Lookup the catalog set
+	auto &catalog_set = duck_schema_entry.GetCatalogSet(type);
+
+	// Use the index to find the actual entry
+	auto entry = catalog_set.GetEntry(transaction, name);
+	return entry;
+}
+
+void DependencyManager::CleanupDependencies(CatalogTransaction transaction, CatalogEntry &object) {
```

Comment:
This also doesn't clean up the DependencySetCatalogEntry, not sure if that's something we want to do?

Probably yes